lib/index.ts

```diff
@@ -31,64 +32,82 @@ export * from './local-image';
 export function splitBuildStream(
 	composition: Compose.Composition,
 	buildStream: Stream.Readable,
-): Promise<BuildTask[]> {
+): Bluebird<BuildTask[]> {
 	const images = Compose.parse(composition);
 	return fromImageDescriptors(images, buildStream);
 }
 
-export function fromImageDescriptors(
-	images: Compose.ImageDescriptor[],
-	buildStream: Stream.Readable,
-): Promise<BuildTask[]> {
+function setDefault(obj: { [s: string]: any }, name: string | number, defaultValue: any): any {
+	// Like Python's `dict.setdefault`
+	const value = obj[name];
+	if (_.isUndefined(value)) {
+		obj[name] = defaultValue;
+		return defaultValue;
+	}
+	return value;
+}
 
-	return new Promise((resolve, reject) => {
+function getAlternateDockerfiles(tasks: BuildTask[]): { [s: string]: BuildTask[] } {
+	const result: { [s: string]: BuildTask[] } = {};
+	for (const task of tasks) {
+		if (task.dockerfile) {
+			setDefault(result, task.dockerfile, []).push(task);
+			task.dockerfile = ALTERNATE_DOCKERFILE_PATH;
+		}
+	}
+	return result;
+}
+
+function copyToTasksBuildStreams(tasks: BuildTask[], buffer: Buffer, header: tar.TarHeader, name: string) {
+	// Adds the file to all tasks build streams as name.
+	const updatedHeader = _.merge(header, { name });
+	tasks.forEach((task) => {
+		task.buildStream!.entry(updatedHeader, buffer);
+	});
+}
+
+export function fromImageDescriptors(images: Compose.ImageDescriptor[], buildStream: Stream.Readable): Bluebird<BuildTask[]> {
+	return new Bluebird((resolve, reject) => {
 		// Firstly create a list of BuildTasks based on the composition
 		const tasks = Utils.generateBuildTasks(images);
 
+		// Dict of { filename: BuildTask } for all alternate Dockerfiles
+		const alternateDockerfiles = getAlternateDockerfiles(tasks);
+
 		const extract = tar.extract();
 
-		const entryFn = (
-			header: tar.TarHeader,
-			stream: Stream.Readable,
-			next: () => void,
-		): void => {
+		extract.on('entry', async (header: tar.TarHeader, stream: Stream.Readable, next: () => void): Promise<void> => {
```

Comment:
Is typescript adding shims for async/await or compiling them verbatim? I mean, async/await is only in fairly recent Node versions right?

Reply:
It depends on the target js version, see https://blog.mariusschulz.com/2016/12/09/typescript-2-1-async-await-for-es3-es5 .
In our case (which is es5 https://github.com/resin-io-modules/resin-multibuild/blob/master/tsconfig.json#L4 ), it will get transpiled. Maybe we should update the target js version as this code will only run in recent node versions?

Reply:
Bear in mind that await/async has pretty terrible performance in nodejs 6 (and early nodejs 8 versions) @zvin because it disables the v8 optimiser, so we should only really be using the native version if we can rely on nodejs 8.3+ or the code has no need to consider performance at all

Reply:
Yeah @zvin I'd rather we either update the builder to use a later version of node, or we don't use async/await

Reply:
@CameronDiver `resin-builder` uses `resin-resin-base:3.3.0` so it has node `8.10.0` https://github.com/resin-io/resin-base/blob/master/Dockerfile#L41 .
What do you think of updating the js target of typescript to `es2017` for this project, `resin-builder` and other things `resin-builder` uses?

Reply:
I was not aware that had been updated. In principle that's fine with me, I'd like @Page- 's take on this though.